### PR TITLE
Fix Star Wars whenGetAll handlers

### PR DESCRIPTION
**Pull request by islemaster**
Only count non-fading items when checking whether we just collected the last item.  Explanation of fix inline in diff.

**Fixes [Pivotal #110267384](https://www.pivotaltracker.com/story/show/110267384)**

### CODE REVIEW

File: apps/src/studio/studio.js [hunk 1]
```diff
@@ -1183,17 +1183,19 @@ function edgeCollidableCollisionDistance (collidable, edgeName, yAxis) {
...
-  // Traverse the list in reverse order because we may remove elements from the
-  // list while inside the loop:
-  for (var i = list.length - 1; i >= 0; i--) {
-    var collidable = list[i];
-    var next = collidable.getNextPosition();
+  var collidable, next, className, numItemsOfClass;
+  var distanceScaling = constants.SPRITE_COLLIDE_DISTANCE_SCALING;
 
-    if (collidable.isFading && collidable.isFading()) {
-      continue;
```
Comment (islemaster) [inline apps/src/studio/studio.js:1193]: Here, you'll notice we do one thing right - we skip items that are already fading out while we loop through items checking collisions.  This makes sure you can't collect an item that's already fading out.

Comment (islemaster) [inline apps/src/studio/studio.js:1193]: Here, you'll notice we do one thing right - we skip items that are already fading out while we loop through items checking collisions.  This makes sure you can't collect an item that's already fading out.

File: apps/src/studio/studio.js [hunk 2]
```diff
@@ -1233,20 +1235,17 @@ function handleActorCollisionsWithCollidableList (
             // NOTE: if items are allowed to move outOfBounds(), this may never
             // be called because the last item may not be removed here.
 
-            if (list.length === 1) {
+            if (activeCollidables.length === 1) {
```
Comment (islemaster) [inline apps/src/studio/studio.js:1238]: Unfortunately, down here where we check whether we are collecting the last item, we look at the size of the whole list... which may include items that are currently fading out.  Therefore, the following case breaks:
1. Make a level with two stormtroopers.
2. Collect stormtrooper #1.  This logic runs, but `list.length` is 2 so nothing happens.
3. Collect stormtrooper #2 while the first one is still fading out.  This logic runs, but `list.length` is still 2, so nothing happens.
4. Both stormtroopers fade out, but the `whenGetAllItems` handlers never runs.

File: apps/src/studio/studio.js [hunk 3]
```diff
@@ -1233,20 +1235,17 @@ function handleActorCollisionsWithCollidableList (
...
               callHandler('whenGetAllItems');
               Studio.trackedBehavior.gotAllItems = true;
             }
 
-            var className = collidable.className;
-            var itemCount = 0;
-            for (var j = 0; j < list.length; j++) {
-              if (className === list[j].className) {
-                itemCount++;
```
Comment (islemaster) [inline apps/src/studio/studio.js:1245]: You can see the same problem here where we count the items of a particular class - items currently fading out get counted.

File: apps/src/studio/studio.js [hunk 4]
```diff
@@ -1255,8 +1254,9 @@ function handleActorCollisionsWithCollidableList (
             collidable.beginRemoveElement();
           } else {
             collidable.removeElement();
-            list.splice(i, 1);
+            list.splice(list.indexOf(collidable), 1);
           }
+          activeCollidables.splice(i, 1);
```
Comment (islemaster) [inline apps/src/studio/studio.js:1259]: My solution is to make a filtered list of `activeCollidables` at the beginning of this algorithm and do all our checks against that.  One consequence is that when we remove an item, we have to remove it from both `list` (which is a reference to `Studio.items`) and from `activeCollidables` so subsequent checks in the loop are correct.  Note that we remove from `activeCollidables` whether we are removing the item immediately or just starting its fade - either way, it's no longer active.

**Comment by breville** [comment]
lgtm, thanks to the detailed description.  probably worth testing against other play lab level types to make sure all's good there?

**Comment by islemaster** [comment]
Did a quick check of the Minecraft and Gumball scripts, seems like getall behaviors are still good there as expected.

**Pull request merged**